### ISSUE TITLE
[expo-notifications] daily/weekly/yearly trigger inputs should not require the "repeats" parameter

### ISSUE THREAD

**Issue by HodgertA** [expo/expo#30577]
### Summary

I am trying to implement notifications in an Expo app using expo-notifications. My trigger object in scheduleNotificationAsync conforms to the [YearlyTriggerInput](https://docs.expo.dev/versions/latest/sdk/notifications/#yearlytriggerinput), but it not behaving as expected.

For iOS, I am receiving notifications as if it was indexed at 1. For example for the month of February I see the notification if I set the month to 2. According to the documentation and from looking at the actual [source code](https://github.com/expo/expo/blob/e377f0cd22db5cd7feb8e80348cd7064db5429b1/packages/expo-notifications/src/scheduleNotificationAsync.ts#L230) the month index should start at 0, and if I wanted a notification for February I should set the month to 1.

This is what it says in the documentation: Note: all properties are specified in JavaScript Date's ranges.

For Android, I get this error:

[Unhandled promise rejection: Error: Failed to schedule the notification. Trigger of type: calendar is not supported on Android.]

_Note: I also get this error on android for WeeklyTriggerInput_

This is how I am scheduling my notifications

```
Notifications.scheduleNotificationAsync({
  content: {
    title: 'Complete a quiz',
    body: "👋 Don't forget solve a quiz today!",
  },
  trigger: {
    hour: 8,
    minute: 0,
    month: 2,
    day: 12
  },
})
```
I am running expo dev builds on physical devices when I see this behaviour. Am I missing something?

### Link to the related docs page

https://docs.expo.dev/versions/latest/sdk/notifications/#yearlytriggerinput

**Comment by ekremgunes** [comment]
You will get same errors for monthly,yearly maybe weekly. I have the same problem mr @douglowder . "Calendar not supported on android" . @HodgertA  did you solve it?

**Comment by fobos531** [comment]
Not sure but based on the code it seems like the `type` is completely missing from the trigger options which tells the native code what type of the trigger the notifications is using.

**Comment by douglowder** [comment]
Tracking this with another local notifications issue in ENG-13132

**Comment by douglowder** [comment]
I was able to get Android to work without getting a promise rejection by setting the `repeats` property, which is required to be true for some reason. I'll have a look at how best to fix the types so that this is not needed.

Currently debugging to test the indexing issues with the month value and will check the other values as well.

```ts
const schedulePushNotification: () => Promise<string> = async () => {
  const date = new Date();
  const trigger: YearlyTriggerInput = {
    day: date.getDay(),
    month: date.getMonth(),
    hour: date.getHours(),
    minute: date.getMinutes(),
    repeats: true,
  };
  try {
    return await scheduleNotificationAsync({
      content: {
        title: "You've got mail! 📬",
        body: 'Here is the notification body',
        data: { data: 'goes here', test: { test1: 'more data' } },
      },
      trigger,
    });
  } catch (e) {
    return `Error scheduling notification: ${e}`;
  }
};
```

**Comment by ekremgunes** [comment]
Off topic but why is the monthly notification not available in expo? There is no extra logic, when I examine the java codes, everything goes by calculating the next date ? 

_Mr @douglowder ?_

**Comment by ekremgunes** [comment]
waiting...

**Comment by douglowder** [comment]
I've looked at the different trigger input types, and it isn't clear why monthly was never included.

I've created tasks  ENG-13272 and ENG-13273 to address the required repeat parameter type and the missing monthly type.

**Comment by ekremgunes** [comment]
Thank you sir , we are waiting the updates 🚀

**Comment by douglowder** [comment]
I'm back to actively working on this issue.

- I've found a way to not require the `repeats: true` parameter in the trigger inputs.
- I do not see any issue with year values when scheduling yearly notifications. I am seeing an issue with the weekday value when scheduling weekly notifications on Android.

**Comment by HodgertA** [comment]
Thanks Douglas!

I'm a bit busy and taking a break from my project so I haven't had the chance to follow this thread. I'll hopefully take a deeper look and confirm your changes fixed my problem in a couple weeks or after you've solved the issue.

Thanks again!

**Comment by douglowder** [comment]
Tested iOS and I am not seeing any issue with the `YearlyTriggerInput`. As a test, I created this method that schedules a yearly notification that will be sent in the next minute, and then yearly afterwards:

```ts
const schedulePushNotificationYearlyAsync: () => Promise<string> = async () => {
  const date = new Date();
  const trigger: YearlyTriggerInput = {
    channelId: 'testApp',
    day: date.getDate(),
    month: date.getMonth(),
    hour: date.getHours(),
    minute: date.getMinutes() + 1,
  };
  try {
    return await scheduleNotificationAsync({
      content: {
        title: "Hello!",
        body: `Yearly notification scheduled ${date.toLocaleString()}`,
        data: { data: 'goes here', test: { test1: 'more data' } },
      },
      trigger,
    });
  } catch (e) {
    return `Error scheduling notification: ${e}`;
  }
};
```

The native iOS code converts the JS month value (range 0-11) to the iOS month representation (range 1-12): https://github.com/expo/expo/blob/7981734a6c1a872dee0902757bf7f2d53a64505f/packages/expo-notifications/ios/EXNotifications/Notifications/Scheduling/EXNotificationSchedulerModule.m#L194

**Comment by douglowder** [comment]
Update:

- I'm going to go ahead and submit a PR to fix the issue I found with the weekday value on Android.
- Not requiring the `repeats: true` to be part of the input is going to require a significant change to the API, in order to reliably distinguish between `DailyTriggerInput` and `CalendarTriggerInput`. Once that is ready, I'll submit a separate PR.

**Comment by douglowder** [comment]
Example code for weekly notifications, showing conversion of weekday values from JS (where values are 0-6) to Android/iOS (where they are 1-7):

```ts
/**
 * Schedule a weekly notification where the first one will appear 1 minute from now
 */
const schedulePushNotificationWeekly: () => Promise<string> = async () => {
  const date = new Date();
  const trigger: WeeklyTriggerInput = {
    channelId: 'testApp',
    weekday: date.getDay() + 1, // convert JS day of week to native
    hour: date.getHours(),
    minute: date.getMinutes() + 1,
  };
  try {
    return await scheduleNotificationAsync({
      content: {
        title: "You've got mail! 📬",
        body: `Weekly notification scheduled ${date.toLocaleString()}`,
        data: { data: 'goes here', test: { test1: 'more data' } },
      },
      trigger,
    });
  } catch (e) {
    return `Error scheduling notification: ${e}`;
  }
};
```

**Comment by douglowder** [comment]
We have made an API change to fix this issue. Since it is potentially a breaking change for certain input types, this will be rolled out in SDK 52.
